### PR TITLE
Require only one of path and (pathname, query)

### PR DESCRIPTION
**Pull request by fornwall**
Requiring data producers to only write either of

- `path`
- `pathname` and `query`

will make it easier and removes duplication in the format.

The change is transparent to the user of python code since the missing properties are recreated from the others when parsing.

### CODE REVIEW

**Review by ksaaskil — APPROVED** [review]
👍

File: http_types/utils.py [hunk 1]
```diff
@@ -59,10 +67,22 @@ def __init__(self):
...
+            path = obj_copy["pathname"]
+            if "query" in obj_copy:
+                path += "?" + urlencode(obj_copy["query"])
+            obj_copy['path'] = path
+
         if not "pathname" in obj_copy:
             path = obj_copy['path']
             obj_copy['pathname'] = parse_pathname(path)
 
+        if not "query" in obj_copy and "path" in obj_copy:
```
Comment (ksaaskil) [inline http_types/utils.py]: I assume it should be an error if `path` is not in `obj_copy` at this point?

Comment (fornwall) [reply http_types/utils.py]: In the case of no query parameters, I think it's ok to have only `pathname` but no `query` property.

UPDATE: Ah, then `path` has already been added from the `pathname` previously, and we process it needlessly. I've updated it to add this check first.

**Comment by aby2s** [comment]
I would add more paranoic test cases to check all possible combinations of path, pathname, and query. The logic seems fine, but it is complicated, so it is easy to miss some corner cases.
